### PR TITLE
Changing SpacingDsl behavior for let ... in node

### LINKED ISSUE #125: remove line break after expanded let ... in
```nix
# One line, everyone's happy
let self = { }; in self
```

https://github.com/nix-community/nixpkgs-fmt/issues/65#issuecomment-507306029 suggests that we could consider `let ... in` just as an `assert ...;` or a `with ... ;` statement that does not indent the following body. In that case example 2 is inconsistent.

```nix
# input
# This should be okay if `let` is considered like a `with` or `assert` 
let self = { }; in
self
```

```nix
# nix-fmt output
# But it is expanded :-(
let
  self = {};
in
self
```

Overall, it seems that the `let ... in body` is expanded (split across multiple lines as long as it is not written on one line. I think that expanding the `let ... in` part should be independent from expanding the body itself and from decoupling the `in` from the body.

The above example is made even worse by the fact that we do not indent toplevel let bodies (https://github.com/nix-community/nixpkgs-fmt/pull/108). The `self` really looks weird and could be on the same line as the `in`.

May I propose the following patterns ?

```nix
let foo = bar; in foo
```
```nix
let foo = bar; in
foo
```
```nix
let
  foo = bar;
in {
  bar = foo;
}
```

The implicit rule is that when the let+in construct is one one line, it behaves like a `with` or `assert`, and when it is on several lines (expanded form), no line break is enforced after the `in` (which means that line breaks are only enforced for a multi-line body after a single line let+in).

### PR DESCRIPTION
**Pull request by Rizary**
This PR fixes #125

### CODE REVIEW

File: test_data/issue-125.bad.nix [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{foo,bar}:
+let
+foo = let x = y; in 
+z;
```
Comment (zimbatm) [inline test_data/issue-125.bad.nix]:
```suggestion
z; 
# all in one line
bar = let x = 3; in x;
```
let's also test that use-case

**Comment by layus** [comment]
Here are snippets I am not too easy with with this change:

```diff
-  imports = [
-    (mkChangedOptionModule [ "services" "ddclient" "domain" ] [ "services" "ddclient" "domains" ]
-      (config:
-        let value = getAttrFromPath [ "services" "ddclient" "domain" ] config;
-        in if value != "" then [ value ] else []))
-    (mkRemovedOptionModule [ "services" "ddclient" "homeDir" ] "")
-  ];
+    imports = [
+      (
+        mkChangedOptionModule [ "services" "ddclient" "domain" ] [ "services" "ddclient" "domains" ]
+          (
+            config:
+              let
+                value = getAttrFromPath [ "services" "ddclient" "domain" ] config;
+              in
+                if value != "" then [ value ] else []
+          )
+      )
+      (mkRemovedOptionModule [ "services" "ddclient" "homeDir" ] "")
+    ];
```

I would have expected fmt to keep the `if` on the same line as the `in`

**Comment by layus** [comment]
```
-let cfg = config.services.xserver.libinput;
-    xorgBool = v: if v then "on" else "off";
-in {
+let
+  cfg = config.services.xserver.libinput;
+  xorgBool = v: if v then "on" else "off";
+in
+{
```

I am still not used to that extra line break after the in, but this is irrelevant to this PR

**Comment by layus** [comment]
```diff
-      // mapAttrs' (n: v: nameValuePair "${n}.slice"   (sliceToUnit   n v)) cfg.slices
-      // mapAttrs' (n: v: nameValuePair "${n}.socket"  (socketToUnit  n v)) cfg.sockets
-      // mapAttrs' (n: v: nameValuePair "${n}.target"  (targetToUnit  n v)) cfg.targets
-      // mapAttrs' (n: v: nameValuePair "${n}.timer"   (timerToUnit   n v)) cfg.timers
-      // listToAttrs (map
-                   (v: let n = escapeSystemdPath v.where;
-                       in nameValuePair "${n}.mount" (mountToUnit n v)) cfg.mounts)
-      // listToAttrs (map
-                   (v: let n = escapeSystemdPath v.where;
-                       in nameValuePair "${n}.automount" (automountToUnit n v)) cfg.automounts);
+      // mapAttrs' (n: v: nameValuePair "${n}.slice" (sliceToUnit n v)) cfg.slices
+      // mapAttrs' (n: v: nameValuePair "${n}.socket" (socketToUnit n v)) cfg.sockets
+      // mapAttrs' (n: v: nameValuePair "${n}.target" (targetToUnit n v)) cfg.targets
+      // mapAttrs' (n: v: nameValuePair "${n}.timer" (timerToUnit n v)) cfg.timers
+      // listToAttrs (
+        map
+          (
+            v: let
+              n = escapeSystemdPath v.where;
+            in
+              nameValuePair "${n}.mount" (mountToUnit n v)
+          ) cfg.mounts
+      )
+      // listToAttrs (
+        map
+          (
+            v: let
+              n = escapeSystemdPath v.where;
+            in
+              nameValuePair "${n}.automount" (automountToUnit n v)
+          ) cfg.automounts
+      );
```

This makes me wonder if we are on the right track....

But anyway, I did not realize the bug report is only about the `let ... in` part being on the same line. For that specific issue, the PR seems perfect !

Thanks :-)

**Comment by zimbatm** [comment]
There is definitely more work to do.

You seem to be using the `let ... \n in ...` notation which then splits the "header". Did you try moving the `in` keyword to the previous line? Eg:

```
let n = escapeSystemdPath v.where; in
nameValuePair "${n}.automount" (automountToUnit n v)) cfg.automounts);
```

The idea is that it that case it behaves a bit like `assert` and `with`.

**Comment by zimbatm** [comment]
Obviously formatting is a big guessing and taste game :)

**Comment by layus** [comment]
I made tests on existing code. This `let ... in`, assert-like format has a lot of potential, but people will have to get used to it. At the moment, the norm for inline let..in stuff seems to be 

```
let a = "foo";
in some: "use of" a
```

With nixpkgs-fmt, to save space, they will have to get used to 

```
let a = "foo"; in
some: "use of" a
```

a bit more haskell-do-ish :-). They can still keep the old format, but it will be expanded on more lines.

I have the feeling that this change is right :tada:

**Comment by Rizary** [comment]
@layus I'll make another PR to fix some of issue you mentioned, and from then I think I'll make as many as possible to the test_data.

**Pull request merged**